### ISSUE TITLE
GCP project ID: Is it fixed at build time?

### ISSUE THREAD

**Issue by manuelbl** [quarkiverse/quarkus-google-cloud-services#277]
We're working on using separate GCP projects for test and production. But we are getting permission errors indicating that the application in the production environment tries to access the secret manager in the test environment.

It looks as if the GCP project ID is compiled into the application (ueber-jar package). Is this the case? Is this the case for all the options documented in https://quarkiverse.github.io/quarkiverse-docs/quarkus-google-cloud-services/main/#authenticating?

Since we plan to build the application (ueber-jar, native app and possibly Docker container) once and use the same artifact in both the test and production environment, the project should be separate for the built package. What options do we have to achieve this?

**Comment by loicmathieu** [comment]
These should be runtime configuration, secret manager is more complex as it needs to access the configuration early enought to be part of the configuration.

This sounds like a bug

**Comment by manuelbl** [comment]
I've decompiled the generated application and found the GCP project ID in the class `io.quarkus.runtime.generated.Config`. It's inserted into a hash map:

```
      var34.put("quarkus.google.cloud.project-id", "our-gcp-project-id");
```

The key/values from `application.properties` are also inserted into the same hash map as well as many other key/values controlling the logging level.

The project ID must be picked up from the environment at build time. Whatever default GCP project we set before building seems to become the built-in project. That's what we want except it should be picked up at runtime.

**Comment by manuelbl** [comment]
I have upgraded my project to the latest version:

- Quarkus 2.8.1
- quarkus-google-cloud-secret-manager 0.11.0

This resolves the bug. The project ID is no longer part of package but instead taken from the environment (service account) at run-time.

I don't understand what exactly has resolved it. The previous version were:

- Quarkus 2.2.2
- quarkus-google-cloud-secret-manager 0.10.0

**Comment by loicmathieu** [comment]
@manuelbl there was fix on the way the secretmanager extension get the project id, it may fix your issue.

By the way, version 1.0.0 is out (but I just noticed I didn't promote it in Github) you should use it. 
In fact, as the extension is in the Quarkus platform, you should not manage the version by yourself but let the platform do it for you, see https://quarkiverse.github.io/quarkiverse-docs/quarkus-google-cloud-services/main/#dependency-management

**Comment by manuelbl** [comment]
@loicmathieu Thanks for the hint re the Quarkus dependency management. I wasn't aware of that. Surprisingly, the uber-jar size has jumped from 53MB to 73MB.

**Comment by manuelbl** [comment]
I thought the problem was fixed but I've just tricked myself in the deployment. It's still present in the latest version even though I can't find our project ID anywhere in uber-jar. But it clearly picks it up at build-time.

**Comment by loicmathieu** [comment]
Can you please provide a reproducer project and the detailed steps to reproduce the issue ?

**Comment by manuelbl** [comment]
Yes, I can do that. Might take a few days though.

**Comment by manuelbl** [comment]
Here it is already:

**General setup**

Two GCP projects are needed. This description assumes that they are:

- `my-project-test`
- `my-project-prod`

In the `my-project-prod` only, create a secret in the secret manager called `test-cookie` and set its value to a random text.

Finally, an environment with a service account is needed to run the Quarkus application in the `my-project-prod` project. I've used a Compute Engine instance.


**The code**

The code is mainly generated on https://code.quarkus.io/ by selecting *RESTEasy Classic* and *Google Cloud Secret Manager*, including starter code.

The modifications then are:

- delete the tests
- modify `application.properties` and the `GreetingResource` class (see below and in attachment)

*application.properties*:
```
app.cookie=${sm//test-cookie}
quarkus.http.port=9980
```

*GreetingResource.java*:
```
public class GreetingResource {

    @ConfigProperty(name = "app.cookie")
    Optional<String> cookie;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "The cookie is: " + cookie.get() + "\n";
    }
}
```

Note that the project name is not in the source code.


**Building for the happy case**

```
$ gcloud config set project my-project-prod
$ mvnw clean package -Dquarkus.package.type=uber-jar
```

Then transfer the built file `issue-277-1.0.0-SNAPSHOT-runner.jar`  to the execution environment and run it:

```
$ /usr/bin/java -jar issue-277-1.0.0-SNAPSHOT-runner.jar
```

From a separate shell in the same execution environment, test the endpoint:

```
$ wget -qO- http://localhost:9980/hello
The cookie is: Chocolate Cookie
```

That's the expected result: From an environment in the `my-project-prod`, the secret in the same environment is accessed.


**Building for the sad case**

```
$ gcloud config set project my-project-test
$ mvnw clean package -Dquarkus.package.type=uber-jar
```

Then transfer the built file to the execution environment (again in the `my-project-prod` project) and run it:

```
$ /usr/bin/java -jar issue-277-1.0.0-SNAPSHOT-runner.jar
```

From a separate shell in the same execution environment, test the endpoint:

```
$ wget http://localhost:9980/hello
--2022-04-20 19:04:45--  http://localhost:9980/hello
Resolving localhost (localhost)... ::1, 127.0.0.1
Connecting to localhost (localhost)|::1|:9980... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2022-04-20 19:04:45 ERROR 500: Internal Server Error.
```

The request fails with HTTP error 500. The error message from the Quarkus app is attached. The suspicious line is:

```
Caused by: com.google.api.gax.rpc.PermissionDeniedException: io.grpc.StatusRuntimeException: PERMISSION_DENIED: Permission 'secretmanager.versions.access' denied for resource 'projects/my-project-test/secrets/test-cookie/versions/latest' (or it may not exist).
```

and in particular:

```
resource 'projects/my-project-test/secrets/test-cookie/versions/latest' 
```

Note that it refers to the project `my-project-test` even though the project name is not present in the source code and even though the application is executed in a different project. It can only have picked up the project name at build time.




Reproducer project: [issue-277.zip](https://github.com/quarkiverse/quarkus-google-cloud-services/files/8524644/issue-277.zip)

Quarkus application error message: [error.txt](https://github.com/quarkiverse/quarkus-google-cloud-services/files/8524645/error.txt)

**Comment by manuelbl** [comment]
Do your instructions for configuring the project ID apply to run-time, build-time or both?

**Comment by loicmathieu** [comment]
It should be runtime.
That's why I ask to defne the property inside the application.properties and override it at runtime using environement variable or system property.

**Comment by manuelbl** [comment]
From all I can tell, the project ID is compiled into the ueber-jar. In your proposal:

- Will the project ID still be picked up from the build environment and compiled into the ueber-jar, but be overridden at run-time?
- Or can I used an invalid project ID in application.properties at build time and that one will be compiled into the ueber-jar?
- Or can I prevent the project ID from being compiled into the ueber-jar completely?

I'm being rather picky here as it is very relevant for security. We need to understand what's going on so we can prevent any security incidents due to a mix up of project IDs.

**Comment by loicmathieu** [comment]
Everyting inside the application.properties will be included in the package.
And if there is nothing, it will do `ServiceOptions.getDefaultProjectId` at runtime (there was a bug and it was done at build time, I'm working on it).
At runtime, you can override it.

You can use a dumb projectId if you want, I didn't thought about the security implication of having a projectId in the final packaging, but as I said, today `ServiceOptions.getDefaultProjectId` is called at build time which is a bug I'm working on.

**Comment by manuelbl** [comment]
Thanks a lot.

What would be the simplest way to include your version into our project? Can I selectively build the secret-manager?

**Comment by loicmathieu** [comment]
No, you need to checkout the branch and build the all project.
Building is quick, you can skip tests and it'll be a few tens of seconds : mvn clean install -DskipTests

**Comment by manuelbl** [comment]
Thank you very much. That fixes it.

This is without configuring `quarkus.google.cloud.project-id`. I've built it while logged in to one Google project and executed in an instance of a different project, and it has correctly picked up the runtime project. So the project ID is no longer compiled into the jar.

**Comment by loicmathieu** [comment]
Fixed by #286
